src/arraytypes/dictencoding.jl

```diff
@@ -114,19 +114,29 @@ function arrowvector(::DictEncodedType, x, i, nl, fi, de, ded, meta; dictencode:
         # dict encoding doesn't exist yet, so create for 1st time
         if DataAPI.refarray(x) === x
             # need to encode ourselves
-            x = PooledArray(x, encodingtype(length(x)))
+            x = PooledArray(x; signed=true, compress=true)
             inds = DataAPI.refarray(x)
+            pool = DataAPI.refpool(x)
         else
-            inds = copy(DataAPI.refarray(x))
+            pool = DataAPI.refpool(x)
+            refa = DataAPI.refarray(x)
+            inds = copyto!(similar(Vector{signed(eltype(refa))}, length(refa)), refa)
         end
-        # adjust to "offset" instead of index
-        for i = 1:length(inds)
-            @inbounds inds[i] -= 1
-        end
-        pool = DataAPI.refpool(x)
         # horrible hack? yes. better than taking CategoricalArrays dependency? also yes.
         if typeof(pool).name.name == :CategoricalRefPool
-            pool = [get(pool[i]) for i = 1:length(pool)]
+            if eltype(x) >: Missing
+                pool = vcat(missing, pool.pool.levels)
+            else
+                pool = pool.pool.levels
```

Comment:
This would avoid relying on internal fields:
```suggestion
            if eltype(x) >: Missing
                pool = vcat(missing, DataAPI.levels(x))
            else
                pool = DataAPI.levels(x)
```